src/haz3lcore/dynamics/Elaborator.re

```diff
@@ -33,7 +33,15 @@ let cast = (ctx: Ctx.t, mode: Mode.t, self_ty: Typ.t, d: DHExp.t) =>
   | SynFun =>
     switch (self_ty) {
     | Unknown(prov) =>
-      DHExp.cast(d, Unknown(prov), Arrow(Unknown(prov), Unknown(prov)))
```

Comment:
let's add a comment here providing an explicit example of why this case is necessary (eg from the issue)